gqlalchemy/loaders.py

```diff
@@ -0,0 +1,416 @@
+# Copyright (c) 2016-2022 Memgraph Ltd. [https://memgraph.com]
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from . import Memgraph
+from .query_builder import QueryBuilder, Unwind
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from gqlalchemy.models import (
+    MemgraphIndex, 
+    MemgraphTrigger, 
+    TriggerEventObject, 
+    TriggerEventType, 
+    TriggerExecutionPhase
+)
+from pyarrow import fs
+from typing import (
+    List, 
+    Dict, 
+    Any, 
+    Optional,
+    Tuple, 
+    Union
+)
+import pyarrow.dataset as ds
+
+
+
+@dataclass(frozen=True)
+class ForeignKeyMapping:
+    foreign_key: str
+    reference_table: str
+    reference_key: str
+    label: str
+    variables: Optional[Dict[str, str]] = None
+
+
+
+@dataclass(frozen=True)
+class OneToManyMapping:
+    mapping: ForeignKeyMapping
+    from_entity: bool = False
+
+
+
+@dataclass(frozen=True)
+class ManyToManyMapping:
+    mapping1: ForeignKeyMapping
+    mapping2: ForeignKeyMapping
+    label: str
+    from_first: bool = True
+
+
+
+Mapping = Union[List[OneToManyMapping], ManyToManyMapping]
+
+
+
+@dataclass
+class TableMapping:
+    table_name: str
+    mapping: Optional[Mapping] = None
+    indices: Optional[List[str]] = None
+
+
+
+class DataSource(ABC):
+    def __init__(self, file_extension: str) -> None:
+        self._file_extension: str = file_extension
+
+    
+    @abstractmethod
+    def load_data(
+        self,
+        collection_name: str,
+        is_cross_table: bool=False
+    ) -> None:
+        pass
+
+
+
+class S3DataSource(DataSource):
+    def __init__(
+        self, 
+        bucket_name: str,
+        s3_access_key: str,
+        s3_secret_key: str,
+        s3_region: str,
+        file_extension: str,
+        s3_session_token: Optional[str] = None,
+    ) -> None:
+        super().__init__(file_extension=file_extension)
+        self._bucket_name: str = bucket_name
+        self._s3_access_key: str = s3_access_key
+        self._s3_secret_key: str = s3_secret_key
+        self._s3_region: str = s3_region
+        self._s3_session_token: Optional[str] = s3_session_token
+
+    
+    def load_data(
+        self,
+        collection_name: str,
+        is_cross_table: bool = False,
+        columns: Optional[List[str]] = None,
+    ) -> None:
+        s3 = fs.S3FileSystem(
```

Comment:
Does it make sense to put those in some _config property of S3DataSource and then just do fs.S3FileSystem(**config), since we don't use any of those explicitly anywhere else

Reply:
We could, but i like type checking

Reply:
However i will look into it further

Reply:
There is type checking if we leave the arguments as they are in _init_ and then just add to private dictionary, isn't it?

Reply:
I didn't understand the comment in the first place, but You were correct. The handling is now different